### PR TITLE
add attack surface detector, with small changes regarding methods for…

### PR DESCRIPTION
**Pull request by jeremychoi**
This PR closes OWASP/wstg#334 .

- [x] This PR handles the issue and requires no additional PRs.
- [x] You have validated the need for this change.

**What did this PR accomplish?**

- add attack surface detector to identify endpoints
- small changes related to HTTP methods considering the recent usage of other methods for REST API

### CODE REVIEW

File: document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md [hunk 1]
```diff
@@ -73,12 +73,84 @@ user=admin&pass=pass123&debug=true&fromtrustIP=true
 
 Testing for application entry points via a gray-box methodology would consist of everything already identified above with one addition. In cases where there are external sources from which the application receives data and processes it (such as SNMP traps, syslog messages, SMTP, or SOAP messages from other servers) a meeting with the application developers could identify any functions that would accept or expect user input and how they are formatted. For example, the developer could help in understanding how to formulate a correct SOAP request that the application would accept and where the web service resides (if the web service or any other function hasn't already been identified during the black-box testing).
 
+#### OWASP Attack Surface Detector
+
+The Attack Surface Detector(ASD) tool uncovers the endpoints of a web application, the parameters these endpoints accept, and the data type of those parameters. This includes the unlinked endpoints a spider will not be able to find, or optional parameters totally unused in client-side code. It also has the capability to calculate the changes in attack surface between two versions of an application.
```
Comment (kingthorin) [inline document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]:
```suggestion
The Attack Surface Detector (ASD) tool uncovers the endpoints of a web application, the parameters these endpoints accept, and the data type of those parameters. This includes the unlinked endpoints a spider will not be able to find, or optional parameters totally unused in client-side code. It also has the capability to calculate the changes in attack surface between two versions of an application.
```

File: document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md [hunk 2]
```diff
@@ -73,12 +73,84 @@ user=admin&pass=pass123&debug=true&fromtrustIP=true
...
 
+#### OWASP Attack Surface Detector
+
+The Attack Surface Detector (ASD) tool uncovers the endpoints of a web application, the parameters these endpoints accept, and the data type of those parameters. This includes the unlinked endpoints a spider will not be able to find, or optional parameters totally unused in client-side code. It also has the capability to calculate the changes in attack surface between two versions of an application.
+
+The Attack Surface Detector is available as a plugin to both ZAP and Burp Suite, and a Command Line Interface (CLI) tool is also available. The CLI tool exports the attack surface as a JSON output, which can then be used by the ZAP and Burp Suite plugin. This is helpful for cases where the source code is not provided to the penetration tester directly. The CLI tool can also be used for other custom integration where you want to discover an application attack surface or changes in the attack surface.
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: You mean where the source code is actually provided, no? This runs against a source code base.

File: document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md [hunk 3]
```diff
@@ -73,12 +73,84 @@ user=admin&pass=pass123&debug=true&fromtrustIP=true
...
+
+##### How to Use
+
+The CLI jar file is downloadable from [https://github.com/secdec/attack-surface-detector-cli/releases](https://github.com/secdec/attack-surface-detector-cli/releases).
+
+You can run the following command for ASD to identify endpoints from the source code of the target web application.
+
+`java -jar attack-surface-detector-cli-1.3.5.jar railsgoat <source-code-path> [flags]`
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: Remove railsgoat

File: document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md [hunk 4]
```diff
@@ -73,12 +73,84 @@ user=admin&pass=pass123&debug=true&fromtrustIP=true
...
+1/1 projects had endpoints generated
+To enable logging include the -debug argument
+```
+
+You can also generate a JSON output file using the `-json` flag, which can be used by the plugin to both ZAP and Burp Suite. See the following links for more details.
+
+- [Home of ASD Plugin for OWASP ZAP](https://github.com/secdec/attack-surface-detector-zap/wiki)
+- [Home of ASD Plugin for PortSwigger Burp](https://github.com/secdec/attack-surface-detector-burp/wiki)
+
+##### Supported Languages and Frameworks
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: Should we be holding this information in the wiki? This should be part of the documentation of the project, and not our own.

File: document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md (hunk 2 above)
Comment (kingthorin) [reply document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: Ya this could be a bit clearer, the way I read this paragraph was with the implication that a client/customer could provide the details (output) to the analyst if they didn't want to provide the source code itself.

File: document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md (hunk 3 above)
Comment (kingthorin) [reply document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]:
```suggestion
`java -jar attack-surface-detector-cli-1.3.5.jar <source-code-path> [flags]`
```

File: document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md (hunk 4 above)
Comment (kingthorin) [reply document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: That's a good point. If some details needs to be included it probably doesn't need a heading/section unto itself. It could just be something like "ASD supports a wide variety of languages/frameworks, such as: &lt;list&gt;"

File: document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md (hunk 2 above)
Comment (jeremychoi) [reply document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: clarified by adding an example.

File: document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md [hunk 5]
```diff
@@ -138,7 +138,7 @@ You can also generate a JSON output file using the `-json` flag, which can be us
 - [Home of ASD Plugin for OWASP ZAP](https://github.com/secdec/attack-surface-detector-zap/wiki)
 - [Home of ASD Plugin for PortSwigger Burp](https://github.com/secdec/attack-surface-detector-burp/wiki)
 
-##### Supported Languages and Frameworks
+ASD supports a wide variety of languages/frameworks, such as:
```
Comment (ThunderSon) [inline document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: @kingthorin Why are we setting those information in here? If someone wants to know more about ASD, they can go to the tool's page. We don't mention how to run ZAP or Burp. Those are outside the scope of the WSTG. What do you think?
I don't see why we should be maintaining such data, it doesn't add anything to the article to make it more relevant to the reader.

Comment (kingthorin) [reply document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: You're right I guess it should totally go. I kinda saw it being reduced to a single line just for further context to the reader but it doesn't really add much value.

File: document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md [hunk 6]
```diff
@@ -73,12 +73,84 @@ user=admin&pass=pass123&debug=true&fromtrustIP=true
...
+- Java: JSPs, Servlets, Struts, Spring MVC
+- C#: ASP.NET MVC, Web Forms
+- Ruby: Rails
+- Python: Django
+
 ## Tools
 
 - [OWASP Zed Attack Proxy (ZAP)](https://www.zaproxy.org/)
 - [Burp Suite](https://www.portswigger.net/burp/)
-- [Fiddler](https://www.telerik.com/fiddler)
```
Comment (kingthorin) [inline document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: Fiddler is no longer a viable option?

Comment (ThunderSon) [reply document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: I was going through Fiddler after ages. They have created some security add-ons ... This should stay, and added to other testing scenarios. Surprised at first, but that's a great move from their side.

File: document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md (hunk 5 above)
Comment (kingthorin) [reply document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: Done in 984028c

File: document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md (hunk 6 above)
Comment (kingthorin) [reply document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: Done in 984028c

Comment (jeremychoi) [reply document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: No objection in Fiddler staying here, but i thought Fiddler was not as popular as Burp or ZAP these days.

Comment (ThunderSon) [reply document/4_Web_Application_Security_Testing/4.2_Information_Gathering/4.2.6_Identify_Application_Entry_Points_WSTG-INFO-006.md]: True. It is not in our realms, it can be used though. I'd say it's even worth a shot to go over it once more. Our audience could be developers learning how to test, and they'd use fiddler much more than Burp or Zap if it's part of their workflow

**Comment by ThunderSon** [comment]
Thanks, @jeremychoi !
